src/charm.py

```diff
@@ -0,0 +1,595 @@
+            self.unit.status = BlockedStatus("Failed to install dependencies")
+            return
+
+        runner_manager = self._get_runner_manager()
+        if runner_manager:
+            self.unit.status = MaintenanceStatus("Downloading runner binary")
+            try:
+                runner_info = runner_manager.get_latest_runner_bin_url()
+                logger.info(
+                    "Downloading %s from: %s", runner_info.filename, runner_info.download_url
+                )
+                self._stored.runner_bin_url = runner_info.download_url
+                runner_manager.update_runner_bin(runner_info)
+            # Safe guard against transient unexpected error.
+            except Exception as err:  # pylint: disable=broad-exception-caught
+                logger.exception("Failed to update runner binary")
+                # Failure to download runner binary is a transient error.
+                # The charm automatically update runner binary on a schedule.
+                self.unit.status = MaintenanceStatus(f"Failed to update runner binary: {err}")
+                return
+            self.unit.status = MaintenanceStatus("Starting runners")
+            try:
+                self._reconcile_runners(runner_manager)
+                self.unit.status = ActiveStatus()
+            except RunnerError as err:
+                logger.exception("Failed to start runners")
+                self.unit.status = MaintenanceStatus(f"Failed to start runners: {err}")
+        else:
+            self.unit.status = BlockedStatus("Missing token or org/repo path config")
+
+    @catch_unexpected_charm_errors
+    def _on_upgrade_charm(self, _event: UpgradeCharmEvent) -> None:
+        """Handle the update of charm.
+
+        Args:
+            event: Event of charm upgrade.
+        """
+        logger.info("Reinstalling dependencies...")
+        self._install_deps()
+        self._start_services()
+
+        logger.info("Flushing the runners...")
+        runner_manager = self._get_runner_manager()
+        if not runner_manager:
+            return
+
+        runner_manager.flush()
+        self._reconcile_runners(runner_manager)
+
+    @catch_unexpected_charm_errors
+    def _on_config_changed(self, _event: ConfigChangedEvent) -> None:
+        """Handle the configuration change.
+
+        Args:
+            event: Event of configuration change.
+        """
+        try:
+            self._event_timer.ensure_event_timer(
+                "update-runner-bin", self.config["update-interval"]
+            )
+            self._event_timer.ensure_event_timer(
+                "reconcile-runners", self.config["reconcile-interval"]
+            )
+        except TimerEnableError as ex:
+            logger.exception("Failed to start the event timer")
+            self.unit.status = BlockedStatus(
+                f"Failed to start timer for regular reconciliation and binary update checks: {ex}"
+            )
+
```

Comment:
Out of curiosity, if the exception is caught, should it return after setting BlockedStatus?

Reply:
I'm guessing this is to trigger the next event, execution of the current event should probably proceed anyway. Perhaps the blocked state should say what the operator can/ should do to resolve?

Reply:
The `return` should be added.
In terms of the block message, I am not sure what the operator can do there. 
Currently, changing the config again to trigger this hook, would retry the installing the timer. However, this does not seem to be a good design.

Maybe an action to manually install the timer, or an automatic retry for this can be added.
Currently, I am thinking not catching the TimerEnableError which causes ErrorStatus and juju would be retrying this event.

Might need to think more on this.

Reply:
Major comments